src/main/java/io.cdap.plugin/Flatten.java

```diff
@@ -0,0 +1,511 @@
+    Schema outputSchema = generateOutputSchema(inputSchema, outputFieldMap.values());
+    pipelineConfigurer.getStageConfigurer().setOutputSchema(outputSchema);
+  }
+
+  private void checkMaxLimitForNestedLevels(Integer levelToLimitFlattening, FailureCollector collector) {
+    if (levelToLimitFlattening != null && levelToLimitFlattening > MAX_NESTED_LEVEL) {
+      collector.addFailure(String.format("Maximum level for nested is %d.", MAX_NESTED_LEVEL), null);
+      collector.getOrThrowException();
+    }
+  }
+
+  private void checkSchemaType(Schema inputSchema, List<String> fieldsToFlatten, FailureCollector collector) {
+    for (String field : fieldsToFlatten) {
+      Schema schemaField = inputSchema.getField(field).getSchema();
+      if (!isRecord(schemaField)) {
+        collector.addFailure(String.format("'%s' cannot be flattened.", field),
+                             "Only fields with schema type of `record` can be flattened.");
+      }
+    }
+  }
+
+  @Override
+  public void initialize(TransformContext context) throws Exception {
+    super.initialize(context);
+
+    failureCollector = context.getFailureCollector();
+
+    List<String> fieldsToFlatten = config.getFieldsToFlatten();
+    if (fieldsToFlatten == null || fieldsToFlatten.isEmpty()) {
+      failureCollector.addFailure("Atleast one field should be selected.", "");
+      failureCollector.getOrThrowException();
+    }
+
+    Schema inputSchema = context.getInputSchema();
+    if (inputSchema == null) {
+      return;
+    }
+
+    checkMaxLimitForNestedLevels(config.getLevelToLimitFlattening(), failureCollector);
+    List<OutputFieldInfo> inputOutputMapping = createOutputFieldsInfo(inputSchema, fieldsToFlatten,
+                                                                      config.getLevelToLimitFlattening());
+    this.inputOutputMapping = handleDuplicateFieldName(inputOutputMapping, config.getPrefix(), failureCollector);
+    this.outputSchema = generateOutputSchema(inputSchema, inputOutputMapping);
+  }
+
+
+  @Override
+  public void transform(StructuredRecord input, Emitter<StructuredRecord> emitter) throws Exception {
+    if (outputSchema == null) {
+      checkMaxLimitForNestedLevels(config.getLevelToLimitFlattening(), failureCollector);
+      List<OutputFieldInfo> inputOutputMapping = createOutputFieldsInfo(input.getSchema(), config.getFieldsToFlatten(),
+                                                                        config.getLevelToLimitFlattening());
+      this.inputOutputMapping = handleDuplicateFieldName(inputOutputMapping, config.getPrefix(), failureCollector);
+      this.outputSchema = generateOutputSchema(input.getSchema(), inputOutputMapping);
+    }
+    StructuredRecord.Builder builder = StructuredRecord.builder(outputSchema);
+
+    for (Schema.Field field : outputSchema.getFields()) {
+      OutputFieldInfo outputFieldInfo = inputOutputMapping.get(field.getName());
```

Comment:
Does this code compile? How are you accessing inputOutputMapping object declared inside the previous block?

Reply:
Yes it compiles. ``inputOutputMapping`` in line has instance scope, also local variable `inputOutputMapping` is renamed now.